followthemoney/statement/entity.py

```diff
@@ -0,0 +1,407 @@
+                    stmt.id = stmt.generate_key()
+                if stmt.id is None:
+                    stmt.id = stmt.generate_key()
+                yield stmt
+
+    @property
+    def statements(self) -> Generator[Statement, None, None]:
+        """Return all statements for this entity, with extra ID statement."""
+        ids: List[str] = []
+        last_seen: Set[str] = set()
+        first_seen: Set[str] = set()
+        for stmt in self._iter_stmt():
+            yield stmt
+            if stmt.id is not None:
+                ids.append(stmt.id)
+            if stmt.last_seen is not None:
+                last_seen.add(stmt.last_seen)
+            if stmt.first_seen is not None:
+                first_seen.add(stmt.first_seen)
+        if self.id is not None:
+            digest = sha1(self.schema.name.encode("utf-8"))
+            for id in sorted(ids):
+                digest.update(id.encode("utf-8"))
+            checksum = digest.hexdigest()
+            # This is to make the last_change value stable across
+            # serialisation:
+            first = self.last_change or min(first_seen, default=None)
+            yield Statement(
+                canonical_id=self.id,
+                entity_id=self.id,
+                prop=BASE_ID,
+                schema=self.schema.name,
+                value=checksum,
+                dataset=self.dataset.name,
+                first_seen=first,
+                last_seen=max(last_seen, default=None),
+            )
+
+    @property
+    def first_seen(self) -> Optional[str]:
+        seen = (s.first_seen for s in self._iter_stmt() if s.first_seen is not None)
+        return min(seen, default=None)
+
+    @property
+    def last_seen(self) -> Optional[str]:
+        seen = (s.last_seen for s in self._iter_stmt() if s.last_seen is not None)
+        return max(seen, default=None)
+
+    @property
+    def datasets(self) -> Set[str]:
+        datasets: Set[str] = set()
+        for stmt in self._iter_stmt():
+            datasets.add(stmt.dataset)
+        return datasets
+
+    @property
+    def referents(self) -> Set[str]:
+        referents: Set[str] = set(self.extra_referents)
+        for stmt in self._iter_stmt():
+            if stmt.entity_id is not None and stmt.entity_id != self.id:
+                referents.add(stmt.entity_id)
+        return referents
+
+    @property
+    def key_prefix(self) -> Optional[str]:
+        return self.dataset.name
+
+    @key_prefix.setter
+    def key_prefix(self, dataset: str) -> None:
+        raise NotImplementedError()
+
+    @property
+    def caption(self) -> str:
+        """The user-facing label to be used for this entity. This checks a list
+        of properties defined by the schema (caption) and returns the first
+        available value. If no caption is available, return the schema label."""
+        if self._caption is None:
+            self._caption = pick_caption(self)
+        return self._caption
```

Comment:
I’m probably missing something here, but I don’t understand why it’s necessary to override this?

* `StatementEntity` is an `EntityProxy` and inherits the `caption` getter. As `StatementEntity` overrides `get`, shouldn’t that work out of the box?
* The only difference between `pick_caption` and the `caption` getter from `EntityProxy` seems to be that the former caches the result and uses `pick_name` from rigour for name props.

Reply:
So to state my question explicitly: Is there a reason why we can’t have the `caption` getters share the same implementation (maybe except for the caching part if sensible).

Reply:
No. I think medium-term if may be interesting to `pick_caption` to take tuples of `(name, lang?)` so that we can pre-select english names when that metadata is available, but that can still be one implementation.